src/rsa.cpp

```diff
@@ -97,6 +97,8 @@ sp_rsa_key rsa::load_from_file(const std::string &path, password_cb on_password)
 	on_password_wrap wrap(on_password);
 
 	r = PEM_read_RSAPrivateKey(f, nullptr, password_loader, &wrap);
+	fclose(f);
```

Comment:
nice catch!

i think it may be better to do is this way though in case `on_password_wrap` throws an exceptions

```C++
auto fl = std::shared_ptr<FILE>(f, std::fclose);
...
r = PEM_read_RSAPrivateKey(fl.get(), nullptr, password_loader, &wrap);
```

Reply:
I try not to use exception handling if not explicitly said "this function throws an exception". I feel it's better do not rely on exceptions and let on_password_wrap deal with errors itself instead of breaking outer code with exception. Anyway, looks a bit unnatural to handle exceptions around C-style OpenSSL library calls :) 

I was looking for a file object like std::ifstream that closes FILE handle on destroy. I did not found the way to get FILE* from ifsream as required by PEM_read_RSAPrivateKey.

Added simple FileHandle class which closes FILE on destroy.

Reply:
lets go with shared_ptr (unique_ptr would be better but there some issues with custom deleters). reason why it is just one of standard ways to handle such cases. which is rather good comparing to custom solution that needs extra support

Reply:
And second fclose call leads to segfault, why? If I remove second fclose, everything is ok.

> [schmidt@localhost build]$ ./jwtpp_test --gtest_filter=jwtpp.load_rsa_from_file
Running main() from /home/schmidt/src/jwtpp/gtest/googletest/src/gtest_main.cc
Note: Google Test filter = jwtpp.load_rsa_from_file
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from jwtpp
[ RUN      ] jwtpp.load_rsa_from_file
*** glibc detected *** ./jwtpp_test: double free or corruption (!prev): 0x0000000001cff8a0 ***
======= Backtrace: =========
/lib64/libc.so.6[0x3b7fe75f3e]
/lib64/libc.so.6[0x3b7fe78dd0]
/lib64/libc.so.6(fclose+0x14d)[0x3b7fe6648d]
./jwtpp_test(_ZNSt19_Sp_counted_deleterIP8_IO_FILEPFiS1_ESaIvELN9__gnu_cxx12_Lock_policyE2EE10_M_disposeEv+0x2c)[0x4d6114]
./jwtpp_test(_ZNSt16_Sp_counted_baseILN9__gnu_cxx12_Lock_policyE2EE10_M_releaseEv+0x42)[0x4a8072]
./jwtpp_test(_ZNSt14__shared_countILN9__gnu_cxx12_Lock_policyE2EED2Ev+0x27)[0x4a7e7f]
./jwtpp_test(_ZNSt12__shared_ptrI8_IO_FILELN9__gnu_cxx12_Lock_policyE2EED1Ev+0x1c)[0x4d5886]
./jwtpp_test(_ZNSt10shared_ptrI8_IO_FILEED1Ev+0x18)[0x4d58a2]
./jwtpp_test(_ZN5jwtpp3rsa14load_from_fileERKSsSt8functionIFvRSbIcSt11char_traitsIcENS_16secure_allocatorIcEEEiEE+0x18b)[0x4d5687]
./jwtpp_test(_ZN29jwtpp_load_rsa_from_file_Test8TestBodyEv+0x9e)[0x4c8600]
...
Aborted (core dumped)

Reply:
I've found the answer
https://codereview.stackexchange.com/questions/4679/shared-ptr-and-file-for-wrapping-cstdio-update-also-dlfcn-h

> I just realized that it is not allowed to fclose a null pointer.

> Unlike shared_ptr, unique_ptr only invokes the deleter if the pointer is non-zero, so we can simplify the implementation